Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1332 @@
+Next up is `Applicative`. A good intuition behind `Applicative`s is that `pure` represents an identity element, while `<*>` distributes one argument over the other. Think of this like a multiplication, where `0 * a = a * 0 = 0`, while `1 * a = a * 1 = a`.
+
+To illustrate this, observe this table:
+
+| |Left parser failed|Left parser succeeded|
+|-|-|-|
+|Right parser failed|Error from left parser|Error from right parser|
+|Right parser succeeded|Error from left parser|Success|
+
+Note that even if both parsers fail, we only want the error from the first one, as we execute the second one only if the first one succeeded.
+
+This is due to the `Either` part of the parser, where `Left` is like `0` and `Right` is like 1. `pure` represents our `1`, so it should create a parser that always succeeds without consuming any input.
+
+And similarly to `Either`, we want `<*>` to fail in case any of the operands result in a failed parse given some input. This implementation tries to run the left operand, which consumes some input and produces a function as well as the unconsumed input. The rest of the input is given to the right operand, and in case of success, it applies the function to the argument.
+
+```hs
+instance Applicative (Parser i e) where
+  pure a = Parser $ \input -> Right (a, input)
+
+  Parser f <*> Parser p = Parser $ \input ->
+    case f input of
+      Left err -> Left err
+      Right (f', rest) ->
+        case p rest of
+          Left err -> Left err
+          Right (output, rest') -> Right (f' output, rest')
+  -- Or:
+  -- Parser f <*> Parser p = Parser $ \input -> do
+  --   (f', rest) <- f input
+  --   (output, rest') <- p rest
+  --   pure (f' output, rest')
+```
+
+In our `Monad` instance, we set `return = pure` as usual. Our bind is more interesting: we try to run the given parser with the input, and if it succeeds, we give the produced output (of type `a`) to our continuation `k`, which produces a parser that is executed with the remainder of the output.
+
+```hs
+instance Monad (Parser i e) where
+  return = pure
+
+  Parser p >>= k = Parser $ \input ->
+    case p input of
+      Left err -> Left err
+      Right (output, rest) ->
+        let
+          Parser p' = k output
+        in
+        p' rest
+  -- Or:
+  -- Parser p >>= k = Parser $ \input -> do
+  --   (output, rest) <- p input
+  --   runParser (k output) rest
+```
+
+With our `Applicative` and `Monad` instances, we are now able to **parse sequentially**. For example, we defined `char` above to let us parse a single character, but what if we wanted to parse many of them?
+
+```hs
+string :: Eq i => [i] -> Parser i e [i]
+string = traverse char
+-- Written in a more verbose but perhaps clear way:
+-- string :: Eq i => [i] -> Parser i e [i]
+-- string [] = pure []
+-- string (x : xs) = (:) <$> char x <*> string xs
+-- Or even in terms of monad:
+-- string [] = return []
+-- string (x : xs) = do
+--   y <- char x
+--   ys <- string xs
+--   return (y : ys)
+```
```

Comment:
I think, it is worth to write those examples not in comments, but in separate blocks, the article is not in literate haskell format anyway.

Reply:
I've made separate blocks for them.